### PR TITLE
Migrate Existing Feedback for Text Exercises to Text Blocks

### PR DESCRIPTION
**Pull request by jpbernius**
### Checklist
- [ ] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [X] I documented my source code using the JavaDoc / JSDoc style.
- [X] I added integration test cases for the server (Spring) related to the features
- [X] ~I added integration test cases for the client (Jest) related to the features~
- [X] ~I added screenshots/screencast of my UI changes~
- [X] ~I translated all the newly inserted strings~

### Motivation and Context
With the Introduction of Automatic Feedback Suggestions for Text Exercises, we introduced the `TextBlock` Entity. This provides a way of referencing a pice of text by its index. This also circumvents certain problems occurring when referencing identical text twice in a single assessment. Before, text was referenced by simply copying the reference text.

### Description
This PR provides a Migration for converting existing `Feedback` to the new `TextBlock` based representation.

### Steps for Testing

1. Before starting Artemis, find existing text feedback in the database and see the `reference` column in the `feedback` table.
2. Start Artemis.
3. Check again in database to find sha1 hashes instead of text in `feedback` table.
4. Check the `text_block` table to find a those same hashes with start/end index and submission reference.

---

**⚠️ This should not be merged as long as new Feedback is saved in the old format. PR incomming.**

~So far, this PR does not contain JavaDoc comments. However, these changes will be reverted as soon as it was run on the production instance, so we can save those efforts towards more useful documentation efforts like #717.~

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/TextFeedbackMigration.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return feedbackRepository.findAllByResult_Submission_Participation_ExerciseIdIn(findAllTextExerciseIds());
+    }
+
+    @NotNull
+    private List<Feedback> findAllTextFeedbacksNeedingMigration() {
+        return findAllTextFeedbacks().parallelStream().filter(feedback -> feedback.getReference() != null && !textBlockIdPattern.matcher(feedback.getReference()).matches())
+                .collect(Collectors.toList());
+    }
+
+    @PostConstruct()
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/TextFeedbackMigration.java]: Build fails here for missing JavaDoc. See comment in [PR description](https://github.com/ls1intum/Artemis/pull/729#issue-308667539):
> So far, this PR does not contain JavaDoc comments. However, these changes will be reverted as soon as it was run on the production instance, so we can save those efforts towards more useful documentation efforts like #717.

File: src/main/java/de/tum/in/www1/artemis/service/TextFeedbackMigration.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import de.tum.in.www1.artemis.domain.Feedback;
+import de.tum.in.www1.artemis.domain.TextBlock;
+import de.tum.in.www1.artemis.domain.TextExercise;
+import de.tum.in.www1.artemis.domain.TextSubmission;
+import de.tum.in.www1.artemis.repository.FeedbackRepository;
+import de.tum.in.www1.artemis.repository.TextBlockRepository;
+import de.tum.in.www1.artemis.repository.TextExerciseRepository;
+
+@Service
+public class TextFeedbackMigration {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/TextFeedbackMigration.java]: append Service

**Comment by krusche** [comment]
@jpbernius what is the status of this PR? Seems to be outdated...
Should we close it and reopen it if more progress is made?

**Comment by krusche** [comment]
There is no progress since some time, so I close this. Feel free to reopen this PR when more progress is made...

**Pull request closed without merging**